File: Carrel/Carrel/Core/Network/Models/RepositoryFile.swift
import Foundation

struct RepositoryFile: Codable, Identifiable, Equatable {
    let name: String
    let path: String
    let type: FileType
    let size: Int?

    var id: String { path }

    enum FileType: String, Codable {
        case file
        case dir
    }

    var isDirectory: Bool { type == .dir }
    var isTexFile: Bool { name.hasSuffix(".tex") }
    var isPdfFile: Bool { name.hasSuffix(".pdf") }
    var isSelectable: Bool { isTexFile || isPdfFile }
}

// MARK: - Preview Support

extension RepositoryFile {
    static var previewDirectory: RepositoryFile {
        RepositoryFile(name: "chapters", path: "chapters", type: .dir, size: nil)
    }

    static var previewTexFile: RepositoryFile {
        RepositoryFile(name: "main.tex", path: "main.tex", type: .file, size: 12345)
    }

    static var previewPdfFile: RepositoryFile {
        RepositoryFile(name: "paper.pdf", path: "paper.pdf", type: .file, size: 524288)
    }
}

File: CarrelAndroid/app/src/main/java/com/carrel/app/core/network/models/RepositoryFile.kt
package com.carrel.app.core.network.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class RepositoryFile(
    val name: String,
    val path: String,
    val type: FileType,
    // Convex returns numbers as doubles
    val size: Double? = null
) {
    val isDirectory: Boolean get() = type == FileType.DIR
    val isTexFile: Boolean get() = name.endsWith(".tex")
    val isPdfFile: Boolean get() = name.endsWith(".pdf")
    val isSelectable: Boolean get() = isTexFile || isPdfFile
    val sizeInt: Int? get() = size?.toInt()
}

@Serializable
enum class FileType {
    @SerialName("file") FILE,
    @SerialName("dir") DIR
}
